File: src/main/java/com/allaboutjava/chap09/Tutorial_13_Future_Zip.java
package com.allaboutjava.chap09;

import io.vavr.Tuple2;
import io.vavr.concurrent.Future;
import io.vavr.control.Option;

import java.util.function.Function;

import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;
import static io.vavr.Predicates.instanceOf;
import static java.lang.Thread.sleep;

public class Tutorial_13_Future_Zip {

    public static void main(String[] args) throws InterruptedException {

        println("Step 1: Define a method which returns a Future Option");
        Function<String, Future<Option<Integer>>> donutStock = donut -> Future.of(() -> {
            // assume some long running database operation
            sleep(500);
            println("checking donut stock");
            return donut.equals("vanilla donut") ? Some(10) : None();
        });

        println("\nStep 2: Define a method which returns a Future Double for donut price");
        final Future<Double> donutPrice = Future.successful(3.25);

        println("\nStep 3: Zip the values of the first future with the second future");
        final Future<Tuple2<Option<Integer>, Double>> donutStockAndPriceOperation = donutStock.apply("vanilla donut").zip(donutPrice);
        donutStockAndPriceOperation
                .onComplete(response ->
                        Match(response).of(
                                Case($Success($()), results -> run(() -> {
                                    printf("Results %s", results);
                                })),
                                Case($Failure($()), e -> run(() ->
                                        printf("Error processing future operations, error = %s", e.getMessage())))
                        ));

        sleep(3000);
    }
}

File: src/main/scala/com/allaboutscala/chap09/Tutorial_13_Future_Zip.scala
package com.allaboutscala.chap09

import scala.util.{Failure, Success}

object Tutorial_13_Future_Zip extends App {

  println("Step 1: Define a method which returns a Future Option")
  import scala.concurrent.ExecutionContext.Implicits.global
  import scala.concurrent.Future
  def donutStock(donut: String): Future[Option[Int]] = Future {
    println("checking donut stock")
    if(donut == "vanilla donut") Some(10) else None
  }

  println(s"\nStep 2: Define a method which returns a Future Double for donut price")
  def donutPrice(): Future[Double] = Future.successful(3.25)

  println(s"\nStep 3: Zip the values of the first future with the second future")
  val donutStockAndPriceOperation = donutStock("vanilla donut") zip donutPrice()
  donutStockAndPriceOperation.onComplete {
    case Success(results) => println(s"Results $results")
    case Failure(e)       => println(s"Error processing future operations, error = ${e.getMessage}")
  }

  Thread.sleep(3000)
}
